File: output/GPT4o-Mini/xCodeEval_4000/C++/Python/6397ab7a3b055b5a0a45676b47d19bc5.py
import sys
import math
from collections import defaultdict

MOD = int(1e9 + 7)

class SegTree:
    def __init__(self, n, init):
        self.n = n
        self.N = 2 ** math.ceil(math.log2(n))
        self.base = self.N - 1
        self.nodes = [init] * (self.base + self.N)

    def left_of(self, index):
        return -1 if index >= self.base else index * 2 + 1

    def right_of(self, index):
        return -1 if index >= self.base else index * 2 + 2

    def parent_of(self, index):
        return -1 if index == 0 else (index - 1) // 2

    def update_rec(self, s, t, l, r, index, v):
        if s == l and t == r:
            self.nodes[index] = v
        else:
            mid = (l + r) // 2
            index_l = self.left_of(index)
            index_r = self.right_of(index)
            if s < mid < t:
                self.update_rec(s, mid, l, mid, index_l, v)
                self.update_rec(mid, t, mid, r, index_r, v)
            elif s < mid:
                self.update_rec(s, t, l, mid, index_l, v)
            elif mid < t:
                self.update_rec(s, t, mid, r, index_r, v)
            self.nodes[index] = self.nodes[index_l] + self.nodes[index_r]

    def add_rec(self, s, t, l, r, index):
        if s == l and t == r:
            return self.nodes[index]
        else:
            mid = (l + r) // 2
            index_l = self.left_of(index)
            index_r = self.right_of(index)
            v0 = 0
            v1 = 0
            if s < mid < t:
                v0 = self.add_rec(s, mid, l, mid, index_l)
                v1 = self.add_rec(mid, t, mid, r, index_r)
            elif s < mid:
                v0 = self.add_rec(s, t, l, mid, index_l)
            elif mid < t:
                v1 = self.add_rec(s, t, mid, r, index_r)
            return v0 + v1

    def update(self, k, v):
        self.update_rec(k, k + 1, 0, self.N, 0, v)

    def sum(self, s, t):
        return self.add_rec(s, t, 0, self.N, 0)

def lower_binary_search(l, r, v, evalfunc):
    if r - l == 1:
        return l if evalfunc(l, v) else l - 1
    m = (l + r) // 2
    if evalfunc(m, v):
        return lower_binary_search(m, r, v, evalfunc)
    else:
        return lower_binary_search(l, m, v, evalfunc)

class bsargv_t:
    def __init__(self, st, s, sum):
        self.st = st
        self.s = s
        self.sum = sum

def evalfunc(val, v):
    buf = v.st.sum(v.s, val + 1)
    return buf <= v.sum / 2

class modll:
    def __init__(self, init=0):
        self.val = init % MOD
        if self.val < 0:
            self.val += MOD

    def __add__(self, x):
        return modll(self.val + x.val)

    def __sub__(self, x):
        return modll(self.val - x.val)

    def __mul__(self, x):
        return modll(self.val * x.val)

    def __truediv__(self, x):
        return modll(self.val * pow(x.val, MOD - 2, MOD) % MOD)

    def __eq__(self, x):
        return self.val == x.val

    def __ne__(self, x):
        return self.val != x.val

    def __repr__(self):
        return str(self.val)

def make_facts(n):
    facts = [modll(1)]
    for i in range(1, n + 1):
        facts.append(facts[-1] * modll(i))
    return facts

def make_ifacts(n, facts):
    ifacts = [modll(1)]
    for i in range(1, n + 1):
        ifacts.append(ifacts[-1] / modll(i))
    return ifacts

def combination(n, r, facts, ifacts):
    if n >= r >= 0:
        return facts[n] * ifacts[r] * ifacts[n - r]
    return modll(0)

def main():
    input = sys.stdin.read
    data = input().split()
    idx = 0
    n = int(data[idx])
    q = int(data[idx + 1])
    idx += 2
    a = list(map(int, data[idx:idx + n]))
    idx += n

    st = SegTree(n, 0)
    ast = SegTree(n, modll(0))
    lst = SegTree(n, modll(0))
    rst = SegTree(n, modll(0))

    for i in range(n):
        w = int(data[idx])
        st.update(i, w)
        ast.update(i, modll(w) * modll(a[i]))
        lst.update(i, modll(w) * modll(n - 1 - i))
        rst.update(i, modll(w) * modll(i))
        idx += 1

    ans = []
    for _ in range(q):
        x = int(data[idx])
        idx += 1
        if x < 0:
            x += 1
            y = int(data[idx])
            idx += 1
            st.update(-x, y)
            ast.update(-x, modll(y) * modll(a[-x]))
            lst.update(-x, modll(y) * modll(n - 1 - (-x)))
            rst.update(-x, modll(y) * modll(-x))
        else:
            y = int(data[idx])
            idx += 1
            x -= 1
            y -= 1
            if x == y:
                ans.append(modll(0))
            else:
                if y - x == 1:
                    halfid = x
                else:
                    v = bsargv_t(st, x, st.sum(x, y + 1))
                    halfid = lower_binary_search(x + 1, y, v, evalfunc)

                cand0 = st.sum(x, halfid + 1)
                cand1 = st.sum(halfid + 1, y + 1)
                dst = 0
                dstid = 0
                if cand0 <= cand1:
                    dst = a[halfid + 1]
                    dstid = halfid + 1
                else:
                    dst = a[halfid]
                    dstid = halfid

                buf = modll(0)
                buf += modll(dst) * st.sum(x, dstid) - ast.sum(x, dstid) - (lst.sum(x, dstid) - modll(st.sum(x, dstid)) * modll(n - 1 - dstid))
                buf += modll(ast.sum(dstid, y + 1)) - modll(dst) * st.sum(dstid, y + 1) - (rst.sum(dstid, y + 1) - modll(st.sum(dstid, y + 1)) * modll(dstid))
                ans.append(buf)

    for result in ans:
        print(result)

if __name__ == "__main__":
    main()

File: output/GPT4o-Mini/xCodeEval_4000/C++/Go/6397ab7a3b055b5a0a45676b47d19bc5.go
package main

import (
	"fmt"
	"math"
)

const (
	eps = 1e-9
	MOD = 1e9 + 7
)

type SegTree struct {
	n, N, base int
	nodes      []int64
}

func NewSegTree(n int, init int64) *SegTree {
	N := int(math.Pow(2, math.Ceil(math.Log2(float64(n)))))
	base := N - 1
	nodes := make([]int64, base+N)
	for i := range nodes {
		nodes[i] = init
	}
	return &SegTree{n: n, N: N, base: base, nodes: nodes}
}

func (st *SegTree) leftOf(index int) int {
	if index >= st.base {
		return -1
	}
	return index*2 + 1
}

func (st *SegTree) rightOf(index int) int {
	if index >= st.base {
		return -1
	}
	return index*2 + 2
}

func (st *SegTree) parentOf(index int) int {
	if index == 0 {
		return -1
	}
	return (index - 1) >> 1
}

func (st *SegTree) updateRec(s, t, l, r, index int, v int64) {
	if s == l && t == r {
		st.nodes[index] = v
	} else {
		mid := (l + r) / 2
		indexL := st.leftOf(index)
		indexR := st.rightOf(index)
		if s < mid && mid < t {
			st.updateRec(s, mid, l, mid, indexL, v)
			st.updateRec(mid, t, mid, r, indexR, v)
		} else if s < mid {
			st.updateRec(s, t, l, mid, indexL, v)
		} else if mid < t {
			st.updateRec(s, t, mid, r, indexR, v)
		}
		st.nodes[index] = st.nodes[indexL] + st.nodes[indexR]
	}
}

func (st *SegTree) Update(k int, v int64) {
	st.updateRec(k, k+1, 0, st.N, 0, v)
}

func (st *SegTree) addRec(s, t, l, r, index int) int64 {
	if s == l && t == r {
		return st.nodes[index]
	} else {
		mid := (l + r) / 2
		indexL := st.leftOf(index)
		indexR := st.rightOf(index)
		var v0, v1 int64
		if s < mid && mid < t {
			v0 = st.addRec(s, mid, l, mid, indexL)
			v1 = st.addRec(mid, t, mid, r, indexR)
		} else if s < mid {
			v0 = st.addRec(s, t, l, mid, indexL)
		} else if mid < t {
			v1 = st.addRec(s, t, mid, r, indexR)
		}
		return v0 + v1
	}
}

func (st *SegTree) Sum(s, t int) int64 {
	return st.addRec(s, t, 0, st.N, 0)
}

type modll struct {
	val int64
}

func (x modll) Modify(value int64) int64 {
	ret := value % int64(MOD)
	if ret < 0 {
		ret += int64(MOD)
	}
	return ret
}

func (x modll) Add(y modll) modll {
	return modll{x.Modify(x.val + y.val)}
}

func (x modll) Sub(y modll) modll {
	return modll{x.Modify(x.val - y.val)}
}

func (x modll) Mul(y modll) modll {
	return modll{x.Modify(x.val * y.val)}
}

func (x modll) Div(y modll) modll {
	return modll{x.Modify(x.val * x.Inv(y.val))}
}

func (x modll) Inv(value int64) int64 {
	if value == 0 {
		return 1 / value
	} else if value == 1 {
		return 1
	}
	return x.Modify(x.Inv(int64(MOD)%value) * x.Modify(-int64(MOD)/value))
}

func (x modll) GetVal() int64 {
	return x.val
}

func lowerBinarySearch(l, r int64, v bsargv_t, evalfunc func(int64, bsargv_t) bool) int64 {
	if r-l == 1 {
		if evalfunc(l, v) {
			return l
		}
		return l - 1
	}
	m := (l + r) / 2
	if evalfunc(m, v) {
		return lowerBinarySearch(m, r, v, evalfunc)
	}
	return lowerBinarySearch(l, m, v, evalfunc)
}

type bsargv_t struct {
	st  *SegTree
	s   int
	sum int64
}

func evalfunc(val int64, v bsargv_t) bool {
	buf := v.st.Sum(v.s, val+1)
	return buf <= v.sum/2
}

func main() {
	var n, q int
	fmt.Scan(&n, &q)
	a := make([]int64, n)
	for i := 0; i < n; i++ {
		fmt.Scan(&a[i])
	}
	st := NewSegTree(n, 0)
	ast := NewSegTree(n, 0)
	lst := NewSegTree(n, 0)
	rst := NewSegTree(n, 0)

	for i := 0; i < n; i++ {
		var w int64
		fmt.Scan(&w)
		st.Update(i, w)
		ast.Update(i, modll{w}.Mul(modll{a[i]}).val)
		lst.Update(i, modll{w}.Mul(modll{int64(n - 1 - i)}).val)
		rst.Update(i, modll{w}.Mul(modll{int64(i)}).val)
	}

	var ans []modll
	for unused := 0; unused < q; unused++ {
		var x int
		fmt.Scan(&x)
		if x < 0 {
			x++
			var y int64
			fmt.Scan(&y)
			st.Update(-x, y)
			ast.Update(-x, modll{y}.Mul(modll{a[-x]}).val)
			lst.Update(-x, modll{y}.Mul(modll{int64(n - 1 - (-x))}).val)
			rst.Update(-x, modll{y}.Mul(modll{int64(-x)}).val)
		} else {
			var y int
			fmt.Scan(&y)
			x--
			y--
			if x == y {
				ans = append(ans, modll{0})
			} else {
				var halfid int
				if y-x == 1 {
					halfid = x
				} else {
					v := bsargv_t{st, x, st.Sum(x, y+1)}
					halfid = int(lowerBinarySearch(int64(x+1), int64(y), v, evalfunc))
				}
				cand0 := st.Sum(x, halfid+1)
				cand1 := st.Sum(halfid+1, y+1)
				var dst int64
				var dstid int
				if cand0 <= cand1 {
					dst = a[halfid+1]
					dstid = halfid + 1
				} else {
					dst = a[halfid]
					dstid = halfid
				}
				buf := modll{0}
				buf = buf.Add(modll{dst}.Mul(modll{st.Sum(x, dstid)}).Sub(ast.Sum(x, dstid)).Sub(lst.Sum(x, dstid).Sub(modll{st.Sum(x, dstid}.Mul(modll{int64(n - 1 - dstid}).val))))
				buf = buf.Add(ast.Sum(dstid, y+1).Sub(modll{dst}.Mul(modll{st.Sum(dstid, y+1)})).Sub(rst.Sum(dstid, y+1).Sub(modll{st.Sum(dstid, y+1)}.Mul(modll{int64(dstid)})))
				ans = append(ans, buf)
			}
		}
	}
	for _, value := range ans {
		fmt.Println(value.GetVal())
	}
}
